File: LeetCode/Java/src/main/java/org/redquark/tutorials/leetcode/MergeTwoSortedLists.java
package org.redquark.tutorials.leetcode;

/**
 * @author Anirudh Sharma
 */
public class MergeTwoSortedLists {

    private static ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        // Check if ant of the lists are null
        if (l1 == null) {
            return l2;
        }
        if (l2 == null) {
            return l1;
        }
        // Head of the result list
        ListNode head;
        // Pointer for the resultant list
        ListNode temp;
        // Choose head which is smaller of the two lists
        if (l1.val < l2.val) {
            temp = head = new ListNode(l1.val);
            l1 = l1.next;
        } else {
            temp = head = new ListNode(l2.val);
            l2 = l2.next;
        }
        // Loop until any of the list becomes null
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                temp.next = new ListNode(l1.val);
                l1 = l1.next;
            } else {
                temp.next = new ListNode(l2.val);
                l2 = l2.next;
            }
            temp = temp.next;
        }
        // Add all the nodes in l1, if remaining
        while (l1 != null) {
            temp.next = new ListNode(l1.val);
            l1 = l1.next;
            temp = temp.next;
        }
        // Add all the nodes in l2, if remaining
        while (l2 != null) {
            temp.next = new ListNode(l2.val);
            l2 = l2.next;
            temp = temp.next;
        }
        return head;
    }

    public static void main(String[] args) {
        ListNode head1 = new ListNode(1);
        head1.next = new ListNode(2);
        head1.next.next = new ListNode(4);
        ListNode head2 = new ListNode(1);
        head2.next = new ListNode(3);
        head2.next.next = new ListNode(4);
        printList(mergeTwoLists(head1, head2));

        printList(mergeTwoLists(null, null));

        head2 = new ListNode(0);
        printList(mergeTwoLists(null, head2));
    }

    private static void printList(ListNode mergedList) {
        while (mergedList != null) {
            System.out.print(mergedList.val + " ");
            mergedList = mergedList.next;
        }
        System.out.println();
    }

    static class ListNode {
        int val;
        ListNode next;

        ListNode(int val) {
            this.val = val;
        }
    }
}

File: LeetCode/Kotlin/src/main/kotlin/org/redquark/tutorials/leetcode/MergeTwoSortedLists.kt
package org.redquark.tutorials.leetcode

fun mergeTwoLists(l1: ListNode21?, l2: ListNode21?): ListNode21? {
    var list1 = l1
    var list2 = l2
    // Check if ant of the lists are null
    if (list1 == null) {
        return list2
    }
    if (list2 == null) {
        return list1
    }
    // Head of the result list
    val head: ListNode21
    // Pointer for the resultant list
    var temp: ListNode21?
    // Choose head which is smaller of the two lists
    if (list1.`val` < list2.`val`) {
        head = ListNode21(list1.`val`)
        temp = head
        list1 = list1.next
    } else {
        head = ListNode21(list2.`val`)
        temp = head
        list2 = list2.next
    }
    // Loop until any of the list becomes null
    while (list1 != null && list2 != null) {
        if (list1.`val` < list2.`val`) {
            temp!!.next = ListNode21(list1.`val`)
            list1 = list1.next
        } else {
            temp!!.next = ListNode21(list2.`val`)
            list2 = list2.next
        }
        temp = temp.next
    }
    // Add all the nodes in l1, if remaining
    while (list1 != null) {
        temp!!.next = ListNode21(list1.`val`)
        list1 = list1.next
        temp = temp.next
    }
    // Add all the nodes in l2, if remaining
    while (list2 != null) {
        temp!!.next = ListNode21(list2.`val`)
        list2 = list2.next
        temp = temp.next
    }
    return head
}

fun main() {
    val head1 = ListNode21(1)
    head1.next = ListNode21(2)
    head1.next!!.next = ListNode21(4)
    var head2 = ListNode21(1)
    head2.next = ListNode21(3)
    head2.next!!.next = ListNode21(4)
    mergeTwoLists(head1, head2)?.let { printList(it) }
    mergeTwoLists(null, null)?.let { printList(it) }
    head2 = ListNode21(0)
    mergeTwoLists(null, head2)?.let { printList(it) }
}

private fun printList(mergedList: ListNode21) {
    var list: ListNode21? = mergedList
    while (list != null) {
        print(list.`val`.toString() + " ")
        list = list.next
    }
    println()
}

class ListNode21(var `val`: Int) {
    var next: ListNode21? = null
}